### ISSUE TITLE
Receive error on long-lived connections to SB

### ISSUE THREAD

**Issue by jacoscaz** [noodlefrenzy/node-amqp10#146]
Hello.

I've stumbled into the following error:

```
{
  "descriptor":{
    "buffer":[0,0,0,0,0,0,0,29],
    "offset":0
  },
  "condition":{
    "contents":"amqp:illegal-state"
  },
  "description":"operation: R:ATTACH, state: OpenReceived"
}
```

This happens on connections to SB's queues that are up for days (weeks, sometimes) and handle large amount of messages. Unfortunately, I still haven't found a way to reproduce the issue, no matter how many thousands of messages I send. Even more unfortunately, this only happens in a remote environment that makes it quite hard to debug.

I am using v1.0.3 and a custom policy that throttles links according to the state of the system. Any idea of what might be causing this?

Cheers.

**Comment by mbroadst** [comment]
@noodlefrenzy my best guess is that it looks like the link is somehow trying to reattach while already attached? 

@jacoscaz it would be very helpful if you could run a copy of your client with the environment variable `DEBUG=amqp*` set, so we could see what kind of framing is occurring during the problem

**Comment by jacoscaz** [comment]
@mbroadst unfortunately, I cannot. Sorry about that, I'd be more than happy to but that specific environment doesn't allow for that.

**Comment by noodlefrenzy** [comment]
@jacoscaz @mbroadst That's possible - I ran into a similar error yesterday when testing some code for V2. I've reached out to my contact inside ServiceBus to see if they might have any insight into when they throw that error so we can match up against what we do.

@jacoscaz are there gaps of time (> 10 minutes) when you are _not_ sending messages? If so, ServiceBus kicks the link and we re-attach the next time a message is sent, which could potentially lead to some sort of overflow of an internal counter or number of handles on their end. Pure speculation at this point.

**Comment by jacoscaz** [comment]
@noodlefrenzy yes, absolutely. I'd say there could be even hours of no messages being sent - mostly at night. 

@noodlefrenzy @mbroadst thanks for your support. I guess I'll refactor things so as to tear down the client completely and set it back up when this thing happens. Does this sound reasonable?

**Comment by mbroadst** [comment]
@jacoscaz yeah I think you're going to have to do that with 1.0.x unfortunately. @noodlefrenzy is currently working on improvements to handle this scenario more seamlessly in 2.0

**Comment by noodlefrenzy** [comment]
I was going to suggest that exact thing, but felt bad about it :)

I'll let you know if I hear anything back from the ServiceBus team. I'll try and build a burn-in test for 2.0 to make sure this doesn't happen with it - can I get any more information about when it happens? How many days, typically? How many messages total? Gaps between messages?

**Comment by jacoscaz** [comment]
@noodlefrenzy my best guess would be anywhere between 2 to 5 days from starting the system @ 100 - 2k messages per day per link, received in bursts of 10 - 100 through 1 - 5 active links. It's a _wild_ guess, though. I'll try to come up with more precise numbers.

No worries - sometimes the old 'just reboot it' trick still comes handy... :).

**Comment by jacoscaz** [comment]
Quick follow-up @noodlefrenzy @mbroadst:

We've just seen the same error making an appearance after two weeks without glitches. This time, though, it's taken a complete reboot of the VM to get things back to normal. Killing and re-starting the node process anew gave 0 results.

**Comment by noodlefrenzy** [comment]
How the... what the... huh? My code doesn't keep any external state - I have no clue what it's doing that would require a reboot. All I can think of is that it's leaking connections on error and causing the VM to run out of sockets. Maybe you can do a `netstat` right after reboot and one after a couple of days and see if there are a bunch of extra listeners lying around?

**Comment by mbroadst** [comment]
@noodlefrenzy maybe we should look into: https://nodejs.org/api/net.html#net_socket_destroy, for error cases

**Comment by noodlefrenzy** [comment]
@mbroadst Good idea. We should definitely make sure that makes it into v2.

**Comment by jacoscaz** [comment]
@noodlefrenzy: I'm sorry I have come across in the wrong way - I absolutely did not mean to imply that your code was responsible for that. My intent is to gather as much context as possible and share it. 

A check with `netstat` is definitely a good idea and I would gladly do that right now if only I had found a way to reproduce this baffling issue. What you suggest makes sense, although (if I understand correctly) it also implies that something is preventing further connections from being established. I will get back at you as soon as I manage to discover more. Many thanks for your awesome work and very kind support - I'm hopeful we'll be able to start working with the next version of amqp10 in the coming weeks.

**Comment by noodlefrenzy** [comment]
@jacoscaz Hahaha, dude, never worry about insulting me - I've been in this industry a long time and have a thick skin. Plus, I've worked at Amazon so can handle just about anything ;-)

I realized you were just trying to give me more context - I was just genuinely baffled by how it might be happening. All I could think of was that we're trapping resources somehow that live beyond process restart, and sockets seemed like the most likely culprit. My thought with the time-delayed `netstat`s was that you wouldn't need to wait for it to replicate, since I'm guessing it's a slow leak and it'll be visible with any time-lapsed snapshots on the same machine.

**Comment by jacoscaz** [comment]
:)

I will give that a try over the next few days and report back with my
findings. Once again, thanks a lot.

On Tuesday, August 18, 2015, Michael Lanzetta <notifications@github.com>
wrote:

> @jacoscaz <https://github.com/jacoscaz> Hahaha, dude, never worry about
> insulting me - I've been in this industry a long time and have a thick
> skin. Plus, I've worked at Amazon so can handle just about anything ;-)
>
> I realized you were just trying to give me more context - I was just
> genuinely baffled by how it might be happening. All I could think of was
> that we're trapping resources somehow that live beyond process restart, and
> sockets seemed like the most likely culprit. My thought with the
> time-delayed netstats was that you wouldn't need to wait for it to
> replicate, since I'm guessing it's a slow leak and it'll be visible with
> any time-lapsed snapshots on the same machine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/noodlefrenzy/node-amqp10/issues/146#issuecomment-132350867>
> .
>


-- 

--

Jacopo Scazzosi
Developer
http://www.jacoscaz.com

**Comment by jacoscaz** [comment]
Hello @noodlefrenzy, hello @mbroadst. Here's another follow-up on this issue. 

Over the last two weeks we had the same thing happen two times. I've been able to check the machine's state (memory, sockets, fds, ...) every 2 - 3 days or so but never found anything out of the ordinary.

However, we did notice that rebooting the machine from within the machine itself (via SSH) doesn't always solve the problem. A reboot from the Azure's console is often necessary to get the AMQP connections to work.

Any news from your contacts in the Azure team?

**Comment by mbroadst** [comment]
@noodlefrenzy this sounds like the handles aren't being reused, I can't recall if that made it in your last set of changes. Did it?

**Comment by noodlefrenzy** [comment]
@mbroadst yes, the handles-not-being-reused issue should be fixed in master. The issue he's seeing is that it's failing to reconnect even after restarting the node process, which doesn't make sense since we don't keep any state.

@jacoscaz - please email me your service-bus namespace and queue-name(s) to milanz@microsoft.com. I'll forward them on to my contacts on that team and see if they can look at the logs on their end.

**Comment by noodlefrenzy** [comment]
Hey @jacoscaz, two things:

1) Did you ever hear back from the Service Bus folks? I forwarded your message with your details on to them, and then kind of lost track and didn't hear anything back myself.

2) v2 of the library is now out (has been for a little bit) and has a TON of bug fixes. It requires some minor rewrites as we've changed to a promise-based API, but if you're willing to do the work it should be much more stable, and might in fact solve this problem. I can't be sure, since honestly it seems like something on the Azure side, but it's worth considering if you have the time. I've updated the examples/ directory with examples for using the new API, so it should be easy to port.

Let me know what I can do to help, I'm terribly sorry this has dragged on.

**Comment by jacoscaz** [comment]
Hello @noodlefrenzy.

1) Nope, I've never heard from them.

2) Yes, I realized a while ago that it had been published on NPM. Unfortunately, I'm still to be scheduled the necessary time to make the switch.

I *really* appreciate your support and I'm sorry I haven't been able to come back at you sooner. I will update you as soon as I get the chance to work on this.

**Comment by noodlefrenzy** [comment]
Let me know if/what I can do to help when you decide to do the switch - I've updated the examples in the `/examples/` directory, but I'm willing to help however you need it. I'm just sorry you ran into such troubles with it - I really hope you won't see those problems when you switch, v2 has a *lot* of bug fixes, and I think the API is cleaner.

**Comment by jacoscaz** [comment]
Hello @noodlefrenzy and @mbroadst. 

Unfortunately we're yet to switch to V2. However, I believe I have some useful feedback (which might also relate to #152). 

As partly mentioned in my above comment dated Aug 18th, we've been dealing with this extremely frustrating issue by manually rebooting any VM that stops receiving messages. As we usually don't get any error whatsoever, we've set-up email alerts based on the amount of transferred data within the hour so that we can intervene when the situation requires it. Far from ideal, yes, but this is the only thing that works, or at least used to work.

This morning, though, one of our VMs stopped receiving messages and we haven't been able to do anything about it so far. We tried rebooting, we tried leaving it off for a while. As usual, no code changes, no errors, nothing. The system looks ok, sockets look ok, FDs look ok, CPU and memory consumption look ok. Even worse, this happened on a relatively "new" VM. It never happened on VMs that have been running (minus the reboots) for much longer.

Setting up the same exact environment (exactly the same code, exactly the same configuration, exactly the same everything) onto another machine, however, worked perfectly.

Regardless of us stumbling into V1's issue of handles not being reused (as per your comment dated Sep 2nd), this looks more and more like something on the SB / Azure side.

**Comment by noodlefrenzy** [comment]
What the heck?! Is this still using the same credentials you emailed me to my work address previously? If so, I'll try and get the SB team to start looking at it ASAP and see if they can find the logs on their side.

**Comment by noodlefrenzy** [comment]
Two more questions for you:

1) Would you be willing to run in "debug mode" (export DEBUG=amqp*)? If not in production, would you be willing to do so against a different queue and on a dedicated VM maybe so you don't interfere with prod? (the latter might not get us the result we need, but is better than nothing). We'd be looking to see if you were getting a disconnect/reconnect loop that might be causing you to overflow counters.

2) Would you be willing to take a "point release" upgrade (e.g. 1.0.3 to 1.0.4) if we were to try and selectively pull some of our V2 bug-fixes "back in time"?

Thanks!
Mike

**Comment by noodlefrenzy** [comment]
P.S. I just heard back from Xin, and it looks like there was a networking issue in the data center with your queue last night which resulted in a temporary outage. That's likely what caused the issues you saw with your new VM.

**Comment by jacoscaz** [comment]
Hello @noodlefrenzy.

Thank you for getting back at me so quickly. Ok, I understand about the networking issue. Will tell the team. Yes, the credentials are still the same. 

> 1) Would you be willing to run in "debug mode" (export DEBUG=amqp*)?

I can try. It'd help if I had a word or a string to look for in the logs. Is there a specific pattern I can work with?

2) Would you be willing to take a "point release" upgrade (e.g. 1.0.3 to 1.0.4) if we were to try and selectively pull some of our V2 bug-fixes "back in time"?

I'd be more than willing to but I am afraid I do not have the time to do so on my schedule. However, it looks like I'll be able to work on V2 in a week or so. Fingers crossed.

Cheers.

**Comment by noodlefrenzy** [comment]
Woohoo, a switch to V2 would be fantastic :) I think that'd almost certainly help clear up a number of issues. I'll take a look at the code for 1.0.3 (it's been a while) and get you something to grep for in debug mode.

**Comment by noodlefrenzy** [comment]
@jacoscaz I think if you look for the string "Transitioning from" that should give us useful information without disclosing anything from your end - we're just trying to get a handle on how often we're seeing state transitions and why.

Looking for "Processing frame" would give us much more information, but some of it would undoubtedly be sensitive. You could limit it to those containing "DetachFrame" without giving away any details, and that would give us a better insight into whether ServiceBus is severing your connections and why.

**Comment by mbroadst** [comment]
@jacoscaz any further information on this? I'd like to close the issue if it's no longer a problem.
